addons/controllers/cpi/cpiconfig_controller.go

```diff
@@ -0,0 +1,314 @@
+	ownerReference := metav1.OwnerReference{
+		APIVersion: clusterapiv1beta1.GroupVersion.String(),
+		Kind:       cluster.Kind,
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+	r.Log.Info("Ensure CPIConfig has the cluster as owner reference")
+	if !clusterapiutil.HasOwnerRef(cpiConfig.OwnerReferences, ownerReference) {
+		r.Log.Info("Adding owner reference to CPIConfig")
+		cpiConfig.OwnerReferences = clusterapiutil.EnsureOwnerRef(cpiConfig.OwnerReferences, ownerReference)
+	}
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretName(cluster.Name, constants.CPIAddonName),
+			Namespace: cpiConfig.Namespace,
+		},
+		Type: v1.SecretTypeOpaque,
+	}
+	secret.SetOwnerReferences([]metav1.OwnerReference{ownerReference})
+
+	mutateFn := func() error {
+		secret.Data = make(map[string][]byte)
+		cpiConfigSpec, err := r.mapCPIConfigToDataValues(ctx, cpiConfig, cluster)
+		r.Log.Info("Mutate CPIConfig data values", "data values", cpiConfigSpec)
+		if err != nil {
+			r.Log.Error(err, "Error occurs when mutate CPIConfig data values")
+			return err
+		}
+		yamlBytes, err := yaml.Marshal(cpiConfigSpec)
+		if err != nil {
+			r.Log.Error(err, "Error marshaling CPIConfig to Yaml")
+			return err
+		}
+		dataValueBytes := append([]byte(constants.TKGDataValueFormatString), yamlBytes...)
+		secret.Data[constants.TKGDataValueFileName] = dataValueBytes
+		return nil
+	}
+	result, err := controllerutil.CreateOrPatch(ctx, r.Client, secret, mutateFn)
+	if err != nil {
+		r.Log.Error(err, "Error creating or patching CPIConfig data values secret")
+		return ctrl.Result{}, err
+	}
+	r.Log.Info(fmt.Sprintf("Resource '%s' data values secret '%s'", constants.CPIAddonName, result))
+	// update the secret reference in CPIConfig status
+	cpiConfig.Status.SecretRef = v1.SecretReference{
+		Name:      secret.Name,
+		Namespace: secret.Namespace,
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// ReconcileCPIConfigDelete triggers when a CPIConfig is deleting, currently do nothing
+func (r *CPIConfigReconciler) ReconcileCPIConfigDelete(context.Context, ctrl.Request, *cpiv1alpha1.CPIConfig, *clusterapiv1beta1.Cluster) (res ctrl.Result, retErr error) {
+	return ctrl.Result{}, nil
+}
+
+// mapCPIConfigToDataValuesNonParavirtual generates CPI data values for non-paravirtual modes
+// It retrieves TLSThumbprint and vSphere server endpoint from VSphereCluster object.
+// It reads vSphere credentials from the cluster secret.
+// It reads the NSX-T credentials from the NSX-T secret.
+// Other values are filled from the specified fields in the CPIConfig CRD.
+func (r *CPIConfigReconciler) mapCPIConfigToDataValuesNonParavirtual(ctx context.Context, cpiConfig *cpiv1alpha1.CPIConfig, cluster *clusterapiv1beta1.Cluster) (*CPIDataValues, error) {
+	d := &CPIDataValues{}
+	c := cpiConfig.Spec.VSphereCPI.NonParavirtualConfig
+	d.VSphereCPI.Mode = CPINonParavirtualMode
+
+	vsphereCluster := &capvv1beta1.VSphereCluster{}
```

Comment:
We get v1beta1 VSphere Cluster here just for getting some common configurations.
Question for addon team, if we need to infer parameters from VSphereCluster, we need to be careful because we don't know about which VSphereCluster version we are deploying with. Is there a better place to get the vsphere credentials?
Otherwise CPI will probably be aware of all the possible VSphereCluster the current CPI is supported.

Reply:
VSphere Cluster is only exist in management cluster.
This logic will break in Workload cluster. We still need some secret in workload cluster, or passing the credentials directly to workload cluster

Reply:
Looks like this is not a concern any more?